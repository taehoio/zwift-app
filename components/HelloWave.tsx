import { Text } from "react-native";
import { preview } from "react-native-ide";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withSequence,
  withTiming,
} from "react-native-reanimated";

export function HelloWave() {
  const rotationAnimation = useSharedValue(0);

  rotationAnimation.value = withRepeat(
    withSequence(
      withTiming(25, { duration: 150 }),
      withTiming(0, { duration: 150 }),
    ),
    4, // Run the animation 4 times
  );

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [{ rotate: `${rotationAnimation.value}deg` }],
  }));

  return (
    <Animated.View style={animatedStyle}>
      <Text className="-mt-1.5 text-4xl">ðŸ‘‹</Text>
    </Animated.View>
  );
}

preview(<HelloWave />);
