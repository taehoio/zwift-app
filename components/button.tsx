import { forwardRef } from "react";
import { Pressable, Text, View } from "react-native";
import { preview } from "react-native-ide";

import { cn } from "@/styles";

export const Button = forwardRef<View, any>(function Button(
  {
    className,
    children,
    ...props
  }: {
    className?: string;
    children?: React.ReactNode;
    props?: any;
  },
  ref,
) {
  return (
    <Pressable
      className={cn(
        "scale-100 rounded-md bg-primary p-4 opacity-100 transition hover:opacity-90 active:scale-90 active:opacity-90",
        className,
      )}
      ref={ref}
      {...props}
    >
      <Text className="font-bold text-white">{children}</Text>
    </Pressable>
  );
});

preview(<Button>Hello World</Button>);
