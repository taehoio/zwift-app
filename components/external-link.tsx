import { Link } from "expo-router";
import {
  WebBrowserPresentationStyle,
  openBrowserAsync,
} from "expo-web-browser";
import { type ComponentProps } from "react";
import { Platform } from "react-native";

import { useColor } from "@/hooks/useColorScheme";

type Props = Omit<ComponentProps<typeof Link>, "href"> & { href: string };

export function ExternalLink({ href, ...rest }: Props) {
  const backgroundColor = useColor("background");

  return (
    <Link
      target="_blank"
      {...rest}
      href={href}
      onPress={async (event) => {
        if (Platform.OS !== "web") {
          // Prevent the default behavior of linking to the default browser on native.
          event.preventDefault();
          // Open the link in an in-app browser.
          await openBrowserAsync(href, {
            toolbarColor: backgroundColor,
            presentationStyle: WebBrowserPresentationStyle.FULL_SCREEN,
            enableBarCollapsing: true,
          });
        }
      }}
    />
  );
}
