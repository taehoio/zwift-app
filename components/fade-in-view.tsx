import { ReactNode } from "react";
import { Platform } from "react-native";
import Animated, { FadeIn as AnimatedFadeIn } from "react-native-reanimated";

export const FadeInView = ({
  children,
  props,
}: {
  children?: ReactNode;
  props?: any;
}) => {
  const isWeb = Platform.OS === "web";

  return isWeb ? (
    <>{children}</>
  ) : (
    <Animated.View entering={AnimatedFadeIn} {...props}>
      {children}
    </Animated.View>
  );
};
