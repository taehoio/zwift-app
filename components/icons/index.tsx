export * from "@/components/icons/logo";
export * from "@/components/icons/chevron-right-icon";
export * from "@/components/icons/clock-icon";
export * from "@/components/icons/dumbbell-icon";
export * from "@/components/icons/flag-icon";
export * from "@/components/icons/ranking-icon";
export * from "@/components/icons/repeat-icon";
export * from "@/components/icons/user-icon";
export * from "@/components/icons/users-icon";
