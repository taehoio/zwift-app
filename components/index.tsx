export * from "@/components/text";
export * from "@/components/button";
