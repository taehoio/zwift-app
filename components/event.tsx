import { forwardRef } from "react";
import { Pressable, View } from "react-native";

import { Text } from "@/components";
import {
  ChevronRightIcon,
  ClockIcon,
  DumbbellIcon,
  FlagIcon,
  RepeatIcon,
  UsersIcon,
} from "@/components/icons";
import { StartTime } from "@/components/start-time";
import { SubgroupBadges } from "@/components/subgroup-badge";
import { cn } from "@/styles";
import { EventWithRoute } from "@/types/zwift.type";

export const Event = forwardRef<View, any>(function Event(
  {
    className,
    eventWithRoute: er,
    isPressable = true,
    showSubgroups = true,
    showFullEventName = false,
    ...props
  }: {
    className?: string;
    eventWithRoute: EventWithRoute;
    isPressable?: boolean;
    showSubgroups?: boolean;
    showFullEventName?: boolean;
  },
  ref,
) {
  return (
    <Pressable
      ref={ref}
      className={cn(
        "flex flex-row gap-0 bg-background py-2 pl-2 pr-0.5",
        className,
      )}
      {...props}
    >
      <View className="flex min-w-[52px] max-w-[52px]">
        <StartTime date={new Date(er.eventStart)} />
      </View>

      <View className="line-clamp-1 flex-1 flex-grow flex-col">
        <EventName eventWithRoute={er} showFullEventName={showFullEventName} />
        <RouteInfo eventWithRoute={er} />
        {showSubgroups && (
          <SubgroupBadges eventId={er.id} subgroups={er.eventSubgroups} />
        )}
      </View>

      {isPressable && (
        <View className="flex w-6 items-end justify-center">
          <ChevronRightIcon className="text-lg color-primary" />
        </View>
      )}
    </Pressable>
  );
});

const EventName = ({
  className,
  eventWithRoute: er,
  showFullEventName = false,
}: {
  className?: string;
  eventWithRoute: EventWithRoute;
  showFullEventName?: boolean;
}) => (
  <View className={cn("flex-1 flex-grow flex-row gap-1", className)}>
    <View className="flex min-w-3.5 max-w-3.5 items-center justify-center">
      {er.eventType === "RACE" && (
        <FlagIcon className="text-md color-default" />
      )}
      {er.eventType === "GROUP_RIDE" && (
        <UsersIcon className="text-md color-default" />
      )}
      {er.eventType === "GROUP_WORKOUT" && (
        <DumbbellIcon className="text-xs color-default" />
      )}
      {er.eventType === "TIME_TRIAL" && (
        <ClockIcon className="text-md color-default" />
      )}
    </View>
    <Text
      className={cn("flex-1 font-bold leading-tight text-secondary", {
        "line-clamp-1": !showFullEventName,
      })}
    >
      {er.name}
    </Text>
  </View>
);

const RouteInfo = ({
  className,
  eventWithRoute: er,
}: {
  className?: string;
  eventWithRoute: EventWithRoute;
}) => (
  <View className={cn("flex flex-row gap-2 pb-0.5", className)}>
    {er.laps > 0 && (
      <View className="flex flex-row gap-1">
        <View className="flex min-w-3.5 max-w-3.5 items-center justify-center">
          <RepeatIcon className="w-3 color-primary" />
        </View>
        <View className="flex flex-row items-center gap-0.5">
          <Text>{er.laps}</Text>
          <Text className="text-xs">laps</Text>
        </View>

        {er.route && (
          <>
            <View className="flex flex-row items-center">
              <Text>
                {(er.route.lead_in_km + er.route.distance_km * er.laps).toFixed(
                  1,
                )}
              </Text>
              <Text className="text-xs">km</Text>
            </View>

            <View className="flex flex-row items-center">
              <Text>{(er.route.elevation_m * er.laps).toFixed(0)}</Text>
              <Text className="text-xs">m</Text>
            </View>
          </>
        )}
      </View>
    )}

    {er.durationInSeconds > 0 && (
      <View className="flex gap-1.5">
        <View className="flex flex-row items-center gap-0.5">
          <View className="flex min-w-3.5 max-w-3.5 items-center justify-center">
            <ClockIcon className="w-3.5 color-primary" />
          </View>
          <Text className="pl-0.5">{er.durationInSeconds / 60}</Text>
          <Text className="text-xs">mins</Text>
        </View>
      </View>
    )}

    {er.distanceInMeters > 0 && (
      <View className="flex gap-1">
        <View className="flex flex-row items-center">
          <Text>{(er.distanceInMeters / 1000).toFixed(1)}</Text>
          <Text className="text-xs">km</Text>
        </View>
      </View>
    )}

    <Text className="line-clamp-1 flex-grow">{er.route?.name}</Text>
  </View>
);
