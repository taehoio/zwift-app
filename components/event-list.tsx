import { Link } from "expo-router";
import { View } from "react-native";

import { Event } from "@/components/event";
import { cn } from "@/styles";
import { EventWithRoute } from "@/types/zwift.type";

export const EventList = ({
  className,
  eventsWithRoute: events,
}: {
  className?: string;
  eventsWithRoute: EventWithRoute[];
}) => {
  return (
    <View className={cn("flex flex-col", className)}>
      {events.map((eventWithRoute, i) => (
        // <Link key={eventWithRoute.id} href={`/events/${eventWithRoute.id}`} asChild>
        <Link key={eventWithRoute.id} href="/welcome" asChild>
          <Event
            className={cn({ "bg-default-200 dark:bg-default-100": isEven(i) })}
            eventWithRoute={eventWithRoute}
          />
        </Link>
      ))}
    </View>
  );
};

const isEven = (num: number) => num % 2 === 0;
