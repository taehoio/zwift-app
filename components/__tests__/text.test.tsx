import { render, screen } from "@testing-library/react-native";
import React from "react";

import { Text } from "@/components";

it("renders correctly", async () => {
  render(
    <>
      <Text>Hello, World!</Text>

      <Text className="font-regular">Hello, World!</Text>
      <Text className="font-medium">Hello, World!</Text>
      <Text className="font-bold">Hello, World!</Text>
      <Text className="font-black">Hello, World!</Text>

      <Text className="font-serif font-black">Hello, World!</Text>

      <Text className="text-primary">Hello, World!</Text>
    </>,
  );

  expect(screen).toMatchSnapshot();
});
