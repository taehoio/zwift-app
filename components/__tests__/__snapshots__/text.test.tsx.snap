// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
[
  <Text
    className="font-sans text-md text-foreground"
    style={
      {
        "fontFamily": "ZwiftSprint-Regular",
      }
    }
  >
    Hello, World!
  </Text>,
  <Text
    className="text-md text-foreground font-regular"
    style={{}}
  >
    Hello, World!
  </Text>,
  <Text
    className="font-sans text-md text-foreground font-medium"
    style={
      {
        "fontFamily": "ZwiftSprint-Medium",
      }
    }
  >
    Hello, World!
  </Text>,
  <Text
    className="font-sans text-md text-foreground font-bold"
    style={
      {
        "fontFamily": "ZwiftSprint-Bold",
      }
    }
  >
    Hello, World!
  </Text>,
  <Text
    className="font-sans text-md text-foreground font-black"
    style={
      {
        "fontFamily": "ZwiftSprint-Black",
      }
    }
  >
    Hello, World!
  </Text>,
  <Text
    className="text-md text-foreground font-serif font-black"
    style={{}}
  >
    Hello, World!
  </Text>,
  <Text
    className="font-sans text-md text-primary"
    style={
      {
        "fontFamily": "ZwiftSprint-Regular",
      }
    }
  >
    Hello, World!
  </Text>,
]
`;
